### PR TITLE
Use named class for UndoPair.

### PR DESCRIPTION
**Pull request by gkossakowski**
Use specific, named class for UndoPair instead of generic Tuple2. This
makes analysis of heap dumps much easier because profilers let you inspect
memory consumption on per-class basis.

### CODE REVIEW

**Comment by gkossakowski** [comment]
Review by @retronym

File: src/reflect/scala/reflect/internal/tpe/TypeConstraints.scala [hunk 1]
```diff
@@ -16,8 +16,9 @@ private[internal] trait TypeConstraints {
   private lazy val _undoLog = new UndoLog
   def undoLog = _undoLog
 
+  final case class UndoPair(tv: TypeVar, tConstraint: TypeConstraint)
```
Comment (retronym) [inline src/reflect/scala/reflect/internal/tpe/TypeConstraints.scala]: Will this carry around an outer pointer? I think it doesn't due to the fact it is final (which is special cased to allow slim nested classes).

Comment (gkossakowski) [inline src/reflect/scala/reflect/internal/tpe/TypeConstraints.scala]: It does carry an outer pointer, good catch:

```
Greks-MBP:build grek (undoPairs)$ javap -private './quick/classes/reflect/scala/reflect/internal/tpe/TypeConstraints$UndoPair'
Compiled from "TypeConstraints.scala"
public final class scala.reflect.internal.tpe.TypeConstraints$UndoPair extends java.lang.Object implements scala.Product,scala.Serializable{
    private final scala.reflect.internal.Types$TypeVar tv;
    private final scala.reflect.internal.tpe.TypeConstraints$TypeConstraint tConstraint;
    private final scala.reflect.internal.SymbolTable $outer;
```

I'll rework this PR.

**Comment by gkossakowski** [comment]
I moved UndoPair to companion object. It doesn't hold an outer pointer reference anymore.

**Comment by retronym** [comment]
LGTM

**Pull request merged**